File: tools/img_to_fb.py
#!/usr/bin/env python3
# img_to_fb.py — convert image to 8bpp row-major framebuffer .mem
# Output: 8192 hex bytes, one per line, addr = row*128 + col
# Usage: python3 img_to_fb.py input.jpg output.mem

import sys
from PIL import Image

src  = sys.argv[1] if len(sys.argv) > 1 else "test.jpg"
dst  = sys.argv[2] if len(sys.argv) > 2 else "build/fb_image.mem"

img = Image.open(src).convert("L").resize((128, 64), Image.LANCZOS)

with open(dst, "w") as f:
    for row in range(64):
        for col in range(128):
            f.write(f"{img.getpixel((col, row)):02x}\n")

print(f"Written {dst}  (128×64 grayscale, row-major)")

File: src/bin/img_to_fb.rs
// img_to_fb — convert image to 8bpp grayscale framebuffer .mem
//
// Output: 8192 hex bytes, one per line, row-major order.
// addr = row*128 + col  (row 0 = top, col 0 = left)
// Used by top_oled_image.v — TRNG hardware dithering compares each
// byte against a random threshold at runtime to produce 1bpp output.
//
// Usage: img_to_fb <input_image> <output.mem>

use image::imageops;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: img_to_fb <input_image> <output.mem>");
        std::process::exit(1);
    }

    let img = image::open(&args[1]).expect("failed to open image");
    // Center-crop to 128×64, preserving aspect ratio (no squish)
    let (sw, sh) = (img.width(), img.height());
    let scale = (128.0 / sw as f64).max(64.0 / sh as f64);
    let nw = (sw as f64 * scale).ceil() as u32;
    let nh = (sh as f64 * scale).ceil() as u32;
    let img = img.resize_exact(nw, nh, imageops::FilterType::Lanczos3);
    let img = image::DynamicImage::from(
        imageops::crop_imm(&img, (nw - 128) / 2, (nh - 64) / 2, 128, 64).to_image()
    );
    let luma = img.to_luma8();

    let hex: String = luma.pixels().map(|p| format!("{:02x}\n", p[0])).collect();
    std::fs::write(&args[2], &hex).expect("failed to write");
    eprintln!("Written {}  (128×64 grayscale, row-major)", args[2]);
}
